src/MongoDB.Driver.Core/Core/Clusters/IClusterExtensions.cs

```diff
@@ -0,0 +1,89 @@
+/* Copyright 2019–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Bindings;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Servers;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    /// <summary>
+    /// This class contains extension methods for ICluster used for server selection with sharded transactions.
+    /// </summary>
+    internal static class IClusterExtensions
+    {
+        public static IServer SelectServerAndPinIfNeeded(
+            this ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            if (TryGetOrClearPinnedShardRouter(cluster, session, out var pinnedServer))
+            {
+                return pinnedServer;
+            }
+
+            // Server selection also updates the cluster type, allowing us to to determine if the shard router
+            // should be pinned.
+            var server = cluster.SelectServer(selector, cancellationToken);
+            PinShardRouterIfNeeded(cluster, session, server);
+            return server;
+        }
+        
+        public static async Task<IServer> SelectServerAndPinIfNeededAsync(
+            this ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            if (TryGetOrClearPinnedShardRouter(cluster, session, out var pinnedServer))
+            {
+                return pinnedServer;
+            }
+
+            // Server selection also updates the cluster type, allowing us to to determine if the shard router
+            // should be pinned.
+            var server = await cluster.SelectServerAsync(selector, cancellationToken).ConfigureAwait(false);
+            PinShardRouterIfNeeded(cluster, session, server);
+
+            return server;
+        }
+        
+        private static void PinShardRouterIfNeeded(ICluster cluster, ICoreSessionHandle session, IServer server)
+        {
+            if (cluster.Description.Type == ClusterType.Sharded && session.IsInTransaction)
+            {
+                session.PinnedShardRouter = server;
+            }
+        }
+
+        private static bool TryGetOrClearPinnedShardRouter(ICluster cluster, ICoreSessionHandle session, out IServer pinnedServer)
```

Comment:
I don't understand this method name.  What part of the implementation is "clearing" anything?